File: 最短网络.py
class DSU: # 并查集
    def __init__(self, n):
        self.pre = list(range(n + 1))
        self.siz = [1 for _ in range(n + 1)]
    
    def find(self, x):
        x, xcopy = self.pre[x], x
        while x != self.pre[x]:
            x = self.pre[x]
        while xcopy != x:
            self.pre[xcopy], xcopy = x, self.pre[xcopy]
        return x
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.siz[x] += self.siz[y]
        self.pre[y] = x
        return True
    
    def size(self, x):
        return self.siz[self.find(x)]
    
n = int(input())
dsu = DSU(n)
edges = []
ans = 0
for i in range(n):
    a = list(map(int, input().split()))
    for j in range(n):
        if a[j]:
            edges.append([a[j], i + 1, j + 1])

edges.sort()
for w, x, y in edges:
    if dsu.merge(x, y):
        ans += w
print(ans)

File: 最短网络.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define int long long

struct DSU { // 并查集
    vector<int> pre, siz;
    
    DSU() {}
    DSU(int n) {
        pre.resize(n);
        std::iota(pre.begin(), pre.end(), 0);
        siz.assign(n, 1);
    }
    
    int find(int x) {
        if (pre[x] == x) {
            return x;
        }
        return pre[x] = find(pre[x]);
    }
    
    bool same(int x, int y) {
        return find(x) == find(y);
    }
    
    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        }
        siz[x] += siz[y];
        pre[y] = x;
        return true;
    }
    
    int size(int x) {
        return siz[find(x)];
    }
};

signed main() {
    int n, ans = 0;
    cin >> n;
    DSU dsu(n + 1);
    vector<pair<int, pair<int, int>>> edges;
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= n; ++j) {
            int w;
            cin >> w;
            if (w) {
                edges.push_back({w, {i, j}});
            }
        }
    }
    sort(edges.begin(), edges.end());

    for (pair<int, pair<int, int>> edge : edges) {
        int w = edge.ff, x = edge.ss.ff, y = edge.ss.ss;
        if (dsu.merge(x, y)) {
            ans += w;
        }
    }
    cout << ans << endl;
    return 0;
}
